text/0028-bee-storage/0022-bee-storage.md

```diff
@@ -0,0 +1,212 @@
+
+    fn insert_transaction(&self, tx: &Tx) -> Result<(), StorageError>;
+    fn find_transaction(&self, tx_hash: TxHash) -> Result<Tx, StorageError>;
+    fn update_transactions_set_solid(
+        &self,
+        transaction_hashes: HashSet<TxHash>,
+    ) -> Result<(), StorageError>;
+    fn update_transactions_set_snapshot_index(
+        &self,
+        transaction_hashes: HashSet<TxHash>,
+        snapshot_index: u32,
+    ) -> Result<(), StorageError>;
+    fn delete_transactions(&self, transaction_hashes: HashSet<TxHash>) -> Result<(), StorageError>;
+
+    //This method is heavy weighted and will be used to populate Tangle struct on initialization
+    fn map_existing_transaction_hashes_to_approvers(
+        &self,
+    ) -> Result<HashesToApprovers, StorageError>;
+    //This method is heavy weighted and will be used to populate Tangle struct on initialization
+    fn map_missing_transaction_hashes_to_approvers(
+        &self,
+    ) -> Result<MissingHashesToRCApprovers, StorageError>;
+
+    //**Operations over milestone's schema**//
+
+    fn insert_milestone(&self, milestone: &Milestone) -> Result<(), StorageError>;
+    fn find_milestone(&self, milestone_hash: MilestoneHash) -> Result<Milestone, StorageError>;
+    fn delete_milestones(
+        &self,
+        milestone_hashes: HashSet<MilestoneHash>,
+    ) -> Result<(), StorageError>;
+
+    //**Operations over state_delta's schema**//
+
+    fn insert_state_delta(
+        &self,
+        state_delta: StateDeltaMap,
+        index: u32,
+    ) -> Result<(), StorageError>;
+
+    fn load_state_delta(&self, index: u32) -> Result<StateDeltaMap, StorageError>;
+}
+
+pub struct Storage<Conn: Connection<Conn>> {
+    connection:   Conn,
+}
+
+impl Storage<DummyConnection> {
+    fn establish_connection(&mut self, url: &str) -> Result<(), ConnectionError> {
+        self.connection = DummyConnection::establish_connection(url)?;
+        Ok(())
+    }
+    fn destroy_connection(connection: DummyConnection) -> Result<(), ConnectionError> {
+        Ok(())
+    }
+}
+
+pub struct DummyConnection {}
+
+impl DummyConnection {
+    fn new() -> Self {
+        Self {}
+    }
+}
+
+impl Connection<DummyConnection> for DummyConnection {
+    fn establish_connection(url: &str) -> Result<DummyConnection, ConnectionError> {
+        Ok(DummyConnection::new())
+    }
+    fn destroy_connection(connection: DummyConnection) -> Result<(), ConnectionError> {
+        Ok(())
+    }
+}
+
+type DummyStorage = Storage<DummyConnection>;
+
+impl Storable for DummyStorage {
+    //Implement all methods here
+}
+
+#[cfg(test)]
+mod tests {
+    #[test]
+    fn it_works() {
+        assert_eq!(2 + 2, 4);
+    }
+}
+
+```
+
+# Drawbacks
+
+There is no reason not to have a storage module
+
+# Rationale and alternatives
+
+- Having storage as a trait allows for multiple backends implementations
+- API for the moment is minimal, but more methods will be added when needed
+
```

Comment:
I think one of alternative could be any established ORM among community.

Reply:
any suggestions for suc an ORM?

Reply:
Ok, i think we can use diesel http://diesel.rs